inventory.lua

```diff
@@ -83,63 +102,106 @@ minetest.register_node("digilines:chest", {
-					sendMessage(pos,"full "..maybeString(stack)..' '..tostring(count))
-				end
-			else
-				-- this happens when the chest has received two stacks in a row and
-				-- filled up exactly with the first one.
-				-- You get a put for the first stack, a put for the second
-				-- and then a overflow with the first in stack and the second in leftover
-				-- and NO full?
-				sendMessage(pos,"overflow "..maybeString(stack)..' '..tostring(count))
+		insert_object = function(pos, _, stack, direction)
+			-- Here, direction = direction item is moving, which is into side.
+			local side = vector.multiply(direction, -1)
+			local leftover = minetest.get_meta(pos):get_inventory():add_item("main", stack)
+			local leftover_count = leftover:get_count()
+			local stack_count = stack:get_count()
+			if leftover_count ~= stack_count then
+				-- Some items were added. Report them.
+				stack:set_count(stack_count - leftover_count)
+				send_message(pos, "tput", stack, side)
+				check_full(pos, stack)
+			end
+			if leftover_count ~= 0 then
+				-- Some items could not be added and bounced back. Report them.
+				send_message(pos, "toverflow", leftover, side)
 			end
 			return leftover
 		end,
+		remove_items = function(pos, _, stack, dir, count)
+			-- Here, stack is the ItemStack in our own inventory that is being
+			-- pulled from, NOT the stack that is actually pulled out.
+			-- Combining it with count gives the stack that is pulled out.
+			-- Also, note that Pipeworks doesn’t pass the index to this
+			-- function, so we use the one recorded in
+			-- allow_metadata_inventory_take; because we don’t implement
+			-- tube.can_remove, Pipeworks will call
+			-- allow_metadata_inventory_take instead and will pass it the
+			-- index.
+			local taken = stack:take_item(count)
+			minetest.get_meta(pos):get_inventory():set_stack("main", last_inventory_take_index, stack)
+			send_message(pos, "ttake", taken, dir)
+			check_empty(pos)
+			return taken
+		end,
 	},
-	allow_metadata_inventory_put = function(pos, _, _, stack)
-		if not can_insert(pos, stack) then
-			sendMessage(pos,"uoverflow "..maybeString(stack))
-		end
+	allow_metadata_inventory_put = function(pos, _, index, stack)
+		-- Remember what was in the target slot before the put; see
+		-- on_metadata_inventory_put for why we care.
+		last_inventory_put_index = index
+		last_inventory_put_stack = minetest.get_meta(pos):get_inventory():get_stack("main", index)
 		return stack:get_count()
 	end,
-	on_metadata_inventory_move = function(pos, _, _, _, _, _, player)
+	allow_metadata_inventory_take = function(_, _, index, stack)
+		-- Remember the index value; see tube.remove_items for why we care.
+		last_inventory_take_index = index
+		return stack:get_count()
+	end,
+	on_metadata_inventory_move = function(pos, from_list, from_index, to_list, to_index, count, player)
 		minetest.log("action", player:get_player_name().." moves stuff in chest at "..minetest.pos_to_string(pos))
 	end,
```

Comment:
Maybe also send here a message.

Reply:
What information would you include in the message? In order to do anything useful with it, I think you would have to include the slot indices the items were moved from and to. But in order to do anything useful with that, I think we would have to include slot indices in all the other messages as well. Should I add those?

Reply:
Yes, that could be important. Eg. if you have (3*8-1)*99 cobble in the chest and want to know if there's enough space for one dirt.

Reply:
So do you think we should have `to` and `from` keys, which are present when applicable? So in a `tput` or `uput` message only `to` would be present, in a `ttake` or `utake` message only `from` would be present, and in a `umove` message `to` and `from` would both be present? These would be the slot numbers that were acted on.

Reply:
This or `slot` where on move it's the diff.
But I think, your way is much better.

Reply:
Fixed in 4d5c32cca7fb. Open to suggestions on the naming of fields in the message, or anything else of course.